elrond-wasm-node/src/api/send_api_node_legacy.rs

```diff
@@ -342,225 +381,281 @@ impl SendApi for ArwenApiImpl {
+            let to_address = to.to_address();
             let _ = executeOnSameContext(
                 gas as i64,
-                address.as_ref().as_ptr(),
+                to_address.as_ptr(),
                 amount_bytes32_ptr,
                 function.as_ptr(),
                 function.len() as i32,
-                arg_buffer.num_args() as i32,
-                arg_buffer.arg_lengths_bytes_ptr(),
-                arg_buffer.arg_data_ptr(),
+                legacy_arg_buffer.num_args() as i32,
+                legacy_arg_buffer.arg_lengths_bytes_ptr(),
+                legacy_arg_buffer.arg_data_ptr(),
+            );
+        }
+    }
+
+    fn execute_on_dest_context_readonly_raw(
+        &self,
+        gas: u64,
+        to: &ManagedAddress<Self::ProxyTypeManager>,
+        endpoint_name: &ManagedBuffer<Self::ProxyTypeManager>,
+        arg_buffer: &ManagedArgBuffer<Self::ProxyTypeManager>,
+    ) -> ManagedVec<Self::ProxyTypeManager, ManagedBuffer<Self::ProxyTypeManager>> {
+        unsafe {
+            let num_return_data_before = getNumReturnData();
+
+            let function = endpoint_name.to_boxed_bytes();
+            let legacy_arg_buffer = arg_buffer.to_legacy_arg_buffer();
+            let to_address = to.to_address();
+            let _ = executeReadOnly(
+                gas as i64,
+                to_address.as_ptr(),
+                function.as_ptr(),
+                function.len() as i32,
+                legacy_arg_buffer.num_args() as i32,
+                legacy_arg_buffer.arg_lengths_bytes_ptr(),
+                legacy_arg_buffer.arg_data_ptr(),
             );
+
+            let num_return_data_after = getNumReturnData();
+            let result_bytes = get_return_data_range(num_return_data_before, num_return_data_after);
+            managed_vec_from_slice_of_boxed_bytes(self.clone(), result_bytes.as_slice())
         }
     }
 
-    fn storage_store_tx_hash_key(&self, data: &[u8]) {
-        let tx_hash = self.get_tx_hash();
-        self.storage_store_slice_u8(tx_hash.as_bytes(), data);
+    fn storage_store_tx_hash_key(&self, data: &ManagedBuffer<Self::ProxyTypeManager>) {
+        let tx_hash = self.get_tx_hash_managed();
+        self.storage_store_managed_buffer_raw(tx_hash.get_raw_handle(), data.get_raw_handle());
     }
 
-    fn storage_load_tx_hash_key(&self) -> BoxedBytes {
-        let tx_hash = self.get_tx_hash();
-        self.storage_load_boxed_bytes(tx_hash.as_bytes())
+    fn storage_load_tx_hash_key(&self) -> ManagedBuffer<Self::ProxyTypeManager> {
+        let tx_hash = self.get_tx_hash_managed();
+        ManagedBuffer::from_raw_handle(
+            self.clone(),
+            self.storage_load_managed_buffer_raw(tx_hash.get_raw_handle()),
+        )
     }
 
     fn call_local_esdt_built_in_function(
         &self,
         gas: u64,
-        function: &[u8],
-        arg_buffer: &ArgBuffer,
-    ) -> Vec<BoxedBytes> {
+        endpoint_name: &ManagedBuffer<Self::ProxyTypeManager>,
```

Comment:
The naming `endpoint_name` is wrong here, as this function calls a built-in function, not a SC endpoint.